File: ch8/Naive_PolyMult/algorithm.cpp
#include <map>
#include <iterator>
#include <iostream>

//
// The functino takes two maps with key as the power of X
// and the corresponding value as the coefficient associated
// with X, and returns a same structure dictionary with the product
// of these two polynomials
//

std::map<int, int> naive_poly_mult(std::map<int, int> P, std::map<int, int> Q) {

    std::map<int, int> M = {};

    for (auto it = P.begin(); it != P.end(); it++) {
        for (auto it2 = Q.begin(); it2 != Q.end(); it2++) {

            int multPower = it->first  + it2->first;
            int multCoeff = it->second * it2->second;
            
            // If the power of X already exist in the product dictionary
            // it add the multCoeff to it, if not it initialize the key and 
            // value to zero, this is how c++ works
            
            M[multPower] += multCoeff;
        }
    }
    
    return M;
}



int main() {
    std::map<int, int> P = {
        {0 , -5},
        {1 , 2},
        {2 , 3}
    };

    std::map<int, int> Q = {
        {0 , 4},
        {1 , -1},
        {2 , 0},
        {3 , 1}
    };

    std::map<int, int> M = naive_poly_mult(P, Q);

    std::cout << "------" << std::endl;
    for (auto it = M.begin(); it != M.end(); it++) {
        std::cout << it->first << " " << it->second << std::endl;
    }
}

File: ch8/Naive_PolyMult/algorithm.cs
using System;
using System.Collections.Generic;   

public class algorithm 
{
    /// <summary>
    /// Takes two Dictionary with the key as the power of X 
    /// and the corresponding value is the coefficient of X
    /// and returns Dictionary with the same structure
    /// </summary>
    /// <param name = "P">The first polynomial</param>
    /// <param name = "Q">The second polynomial</param>
    /// <returns> The product of <paramref name="P"/> and <paramref name="Q"/> </returns>
    public static Dictionary<int, int> naive_poly_mult(Dictionary<int, int> P, Dictionary<int, int> Q) {
        Dictionary<int, int> M = new Dictionary<int, int>();
        foreach(var Element1 in P) {
            foreach(var Element2 in Q) {
                int multPower = Element1.Key + Element2.Key;
                int multCoeff = Element1.Value * Element2.Value;
                if (M.ContainsKey(multPower)) {
                    M[multPower] += multCoeff;
                } else {
                    M.Add(multPower, multCoeff);
                }
            }
        }
        return M;
    }

    public static void Main(string[] args) {
        Dictionary<int, int> P = new Dictionary<int, int>() {
            {0 , -5},
            {1 , 2},
            {2 , 3}
        };

        Dictionary<int, int> Q = new Dictionary<int, int>() {
            {0 , 4},
            {1 , -1},
            {2 , 0},
            {3 , 1}
        };

        var M = naive_poly_mult(P, Q);

        foreach(var ele1 in M) 
        { 
            Console.WriteLine("{0} and {1}", 
                     ele1.Key, ele1.Value); 
        } 
        Console.WriteLine(); 
    } 

}